### DISCUSSION TITLE
How to generate C code and MEX file for a QP calculation program using Matlab

### DISCUSSION THREAD

**Discussion by koumi730** [casadi/casadi#3472] [Q&A]
I would like to efficiently compute QP in Matlab. To achieve this, I need to generate C code from an M-file and create a Mex file. However, I am encountering difficulties. The Matlab version is R2022a.

1. When attempting to generate a Mex file for the Proxqp solver, I encounter an error stating "#error Code generation not supported for ProxqpInterface," preventing the generation of C code. I would appreciate guidance on how to efficiently compute using the proxqp solver in Matlab.

2. When changing the qpprox solver to osqp for C code generation, I encounter an error during Mex file generation indicating the absence of a header file. The error message is 'fatal error: osqp/osqp.h: No such file or directory #include <osqp/osqp.h>.'

```matlab
import casadi.*

% H_val = [1 -1; -1 2]; 
% g_val = [-2; -6];
% A_val = [1 1; -1 2; 2 1];
% uba_val = [2; 2; 3];

%% CasADi
% Getting and setting elements in matrices
% Symbols/expressions
% min.  1/2(x'Hx)+g'x
% s.t.  Ax <= uba
H = MX.sym('H_val',2,2);
g = MX.sym('g_val',2,1); 
A = MX.sym('A_val',3,2); 
uba = MX.sym('uba_val',3,1);

qp = struct; 
qp.h = H.sparsity();
qp.a = A.sparsity();

% Function Definition
% S = conic('S','osqp',qp); 
S = conic('S','proxqp',qp); 
r = S('h', H, 'g', g,'a', A, 'uba', uba); 

%% generated code
opts = struct('main', true,... 
    'mex', true); 
S.generate('test.c',opts);
type('test.c')

mex test.c -DMATLAB_MEX_FILE

``` 

I would like guidance on resolving these two issues."

**Comment by jgillis** [comment]
Is this for an embedded device, or simply on your desktop computer?
In the latter case, and for proxqp, you may use the CasADi C API ( https://github.com/casadi/casadi/blob/main/docs/examples/cplusplus/c_api_usage.cpp )

For the osqp part, see your followup question https://github.com/casadi/casadi/discussions/3502
